tests/run-with-compiler-custom-args/staged-streams_1.scala

```diff
@@ -0,0 +1,697 @@
+          /** Helper function that orchestrates the handling of the function that represents an `advance: Unit => Unit`.
+            * It reifies a nested stream as calls to `advance`. Advance encodes the step function of each nested stream.
+            * It is used in the init of a producer of a nested stream. When an inner stream finishes, the
+            * `nadv` holds the function to the `advance` function of the earlier stream.
+            * `makeAdvanceFunction`, for each nested stream, installs a new `advance` function that after
+            * the stream finishes it will restore the earlier one.
+            *
+            * When `advance` is called the result is consumed in the continuation. Within this continuation
+            * the resulting value should be saved in a variable.
+            *
+            * @param  nadv variable that holds a function that represents the stream at each level.
+            * @param  k              the continuation that consumes a variable.
+            * @return the quote of the orchestrated code that will be executed as
+            */
+          def makeAdvanceFunction[A](nadv: Var[Unit => Unit], k: A => Expr[Unit], stream: StagedStream[A]): Expr[Unit] = {
+            stream match {
+              case Linear(producer) =>
+                producer.card match {
+                  case AtMost1 => producer.init(st => '{
+                    if(~producer.hasNext(st)) {
+                      ~producer.step(st, k)
+                    }
+                    else {
+                      val f = ~nadv.get
+                      f(())
+                    }
+                  })
+                  case Many => producer.init(st => '{
+                    val oldnadv: Unit => Unit = ~nadv.get
+                    val adv1: Unit => Unit = { _: Unit => {
+                      if(~producer.hasNext(st)) {
+                        ~producer.step(st, k)
+                      }
+                      else {
+                        ~nadv.update('{oldnadv})
+                        oldnadv(())
+                      }
+                    }}
+
+                    ~nadv.update('{adv1})
+                    adv1(())
+                  })
+                }
+              case nested: Nested[A, bt] =>
+                makeAdvanceFunction(nadv, (a: bt) => makeAdvanceFunction(nadv, k, nested.nestedf(a)), Linear(nested.producer))
+            }
+          }
+
+          new Producer[Expr[A]] {
+            // _1: if the stream has ended,
+            // _2: the current element,
+            // _3: the step of the inner most steam
+            type St = (Var[Boolean], Var[A], Var[Unit => Unit])
+            val card: Cardinality = Many
+
+            def init(k: St => Expr[Unit]): Expr[Unit] = {
+              producer.init(st =>
+                Var('{ (_: Unit) => ()}){ nadv => {
+                  Var('{ true }) { hasNext => {
+                     Var('{ null.asInstanceOf[A] }) { curr => '{
```

Comment:
I’m wondering if this can cause subtle bugs, since `null.asInstanceOf[Int]` produces `0`, which is undistinguishable from the value `0` (by contrast with references, where `null` can always be distinguished from an object reference).

Why don’t you use `Option[A]` instead?